patients/repo.go

```diff
@@ -931,3 +936,233 @@ func cmpToMongoFilter(cmp *string) (string, bool) {
+	}
+
+	if *params.Period != "1d" && *params.Period != "7d" && *params.Period != "14d" && *params.Period != "30d" {
+		return nil, errors.New("provided period is not one of the valid periods")
+	}
+
+	type Category struct {
+		Heading    string
+		Field      string
+		Comparison string
+		Value      float64
+	}
+
+	categories := [...]Category{
+		{
+			Heading:    "timeInVeryLowPercent",
+			Field:      "timeInVeryLowPercent",
+			Comparison: "$gt",
+			Value:      0.01,
+		},
+		{
+			Heading:    "timeInLowPercent",
+			Field:      "timeInLowPercent",
+			Comparison: "$gt",
+			Value:      0.04,
+		},
+		{
+			Heading:    "dropInTimeInTargetPercent",
+			Field:      "timeInTargetPercentDelta",
+			Comparison: "$lt",
+			Value:      -0.15,
+		},
+		{
+			Heading:    "timeInTargetPercent",
+			Field:      "timeInTargetPercent",
+			Comparison: "$lt",
+			Value:      0.7,
+		},
+		{
+			Heading:    "timeCGMUsePercent",
+			Field:      "timeCGMUsePercent",
+			Comparison: "$lt",
+			Value:      0.7,
+		},
+	}
+
+	limit := 50
+	exclusions := make([]*primitive.ObjectID, 0, 50)
+	tide := Tide{
+		Config: &TideConfig{
+			ClinicId: &clinicId,
+			Filters: &TideFilters{
+				TimeInVeryLowPercent:      ptr(">0.01"),
+				TimeInLowPercent:          ptr(">0.04"),
+				DropInTimeInTargetPercent: ptr("<-0.15"),
+				TimeInTargetPercent:       ptr("<0.7"),
+				TimeCGMUsePercent:         ptr("<0.7"),
+			},
+			HighGlucoseThreshold:     ptr(10.0),
+			LastUploadDateFrom:       params.CgmLastUploadDateFrom,
+			LastUploadDateTo:         params.CgmLastUploadDateTo,
+			LowGlucoseThreshold:      ptr(3.9),
+			Period:                   params.Period,
+			SchemaVersion:            ptr(1),
+			Tags:                     params.Tags,
+			VeryHighGlucoseThreshold: ptr(13.9),
+			VeryLowGlucoseThreshold:  ptr(3.0),
+		},
+		Results: &TideResults{},
+	}
+
+	for _, category := range categories {
+		selector := bson.M{
+			"_id":      bson.M{"$nin": exclusions},
+			"clinicId": clinicObjId,
+			"tags":     bson.M{"$all": tags},
+			"summary.cgmStats.dates.lastUploadDate": bson.M{
+				"$gt":  params.CgmLastUploadDateFrom,
+				"$lte": params.CgmLastUploadDateTo,
+			},
+			"summary.cgmStats.periods." + *params.Period + "." + category.Field: bson.M{category.Comparison: category.Value},
+		}
+
+		opts := options.Find()
+		opts.SetLimit(int64(limit))
+
+		if category.Comparison == "$gt" {
+			opts.SetSort(bson.D{
+				{"summary.cgmStats.periods." + *params.Period + "." + category.Field, -1},
+			})
+		} else {
+			opts.SetSort(bson.D{
+				{"summary.cgmStats.periods." + *params.Period + "." + category.Field, 1},
+			})
+		}
+
+		cursor, err := r.collection.Find(ctx, selector, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		var patientsList []*Patient
+		if err = cursor.All(ctx, &patientsList); err != nil {
+			return nil, fmt.Errorf("error decoding patients list: %w", err)
+		}
+
+		categoryResult := make([]TideResultPatient, 0, 25)
```

Comment:
nitpick: to me, it would seem to make more sense to use a capacity equal to the length of the patientsList rather than an arbritrary number.

Reply:
that makes sense, done.